native/server/src/memory_brongnal.rs

```diff
@@ -1,180 +1,132 @@
+use crate::brongnal::Storage;
 use ed25519_dalek::VerifyingKey;
-use proto::service::brongnal_server::Brongnal;
-use protocol::bundle::verify_bundle;
-use server::parse_verifying_key;
 use server::proto;
 use std::sync::Mutex;
 use std::{collections::HashMap, sync::Arc};
-use tokio::sync::mpsc;
-use tokio::sync::mpsc::Sender;
-use tokio_stream::wrappers::ReceiverStream;
-use tonic::{Request, Response, Status};
+use tonic::Status;
 use x25519_dalek::PublicKey as X25519PublicKey;
 
 #[derive(Clone, Debug)]
-pub struct MemoryBrongnal {
+pub struct MemoryStorage {
     identity_key: Arc<Mutex<HashMap<String, VerifyingKey>>>,
-    current_pre_key: Arc<Mutex<HashMap<String, protocol::x3dh::SignedPreKey>>>,
+    current_pre_key: Arc<Mutex<HashMap<String, proto::service::SignedPreKey>>>,
     one_time_pre_keys: Arc<Mutex<HashMap<String, Vec<X25519PublicKey>>>>,
-    messages: Arc<Mutex<HashMap<String, Vec<protocol::x3dh::Message>>>>,
-    receivers: Arc<Mutex<HashMap<String, Sender<Result<proto::service::Message, Status>>>>>,
+    messages: Arc<Mutex<HashMap<String, Vec<proto::service::Message>>>>,
 }
 
-impl Default for MemoryBrongnal {
-    fn default() -> Self {
-        Self::new()
+impl Storage for MemoryStorage {
+    fn add_user(
+        &self,
+        identity: String,
+        identity_key: VerifyingKey,
+        signed_pre_key: proto::service::SignedPreKey,
+    ) -> tonic::Result<()> {
+        self.identity_key
+            .lock()
+            .unwrap()
+            .insert(identity.clone(), identity_key);
+        self.current_pre_key
+            .lock()
+            .unwrap()
+            .insert(identity.clone(), signed_pre_key);
+        self.one_time_pre_keys
+            .lock()
+            .unwrap()
+            .insert(identity, Vec::new());
```

Comment:
This is probably a language question, but why does identity not need clone here and why are we creating one_time_pre_keys empty with Vec::new()?

Reply:
Identity does not need to be cloned because it is moved in this call. There are no subsequent usages of identity so it is safe to move it here.